include/swift/AST/Pattern.h

```diff
@@ -50,6 +50,17 @@ enum : unsigned { NumPatternKindBits =
 llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, PatternKind kind);
 
 /// Pattern - Base class for all patterns in Swift.
+///
+/// Patterns match a value and binds new variables to parts or all of it
+/// The patterns contain the reference to the variable declaration that is being
+/// assigned to, / or bound, in an assignment.
+/// e.g. In `let (a, _) = (1, 2)`, the pattern is the part after the `let` and
+/// before the `=`. This pattern is a TuplePattern, which contains a
+/// NamedPattern representing the variable declaration of `A` and an AnyPattern.
+/// Patterns also appear in switch cases. In this example `case .some(let foo)`,
+/// where we unwrap an optional in a switch-case, there is an
+/// EnumElementPattern matching the `.some`, a BindingPattern for the `let`,
+/// and a `NamedPattern` referencing the variable declaration of `foo`.
```

Comment:
I feel that that using code would be more helpful in describing the structure compared to using prose.

```
Patterns match a value and optionally create bindings. You can think of
patterns are forming little ASTs similar to expressions. In the language,
patterns can appear in any place where you can create a binding; here
are some examples:

\code
let x : T = ...  // (x : T) → TypedPattern (NamedPattern "x") 
let (a, _) = ... // (a, _) → TuplePattern [NamedPattern "a", AnyPattern]
if let (a, _) = ... { ... } // same as above

for (_, (_, y)) in ["A": (3, 3)] {}
 // (_, (_, y)) → TuplePattern [AnyPattern, TuplePattern [AnyPattern, NamedPattern "y"]]

switch subject {
case .x(0): ()
  // .x(0) → EnumElementPattern "x" (TuplePattern [ExprPattern (LiteralExpr 0)])
case .x(let a): ()
  // .x(let a) → EnumElementPattern "x" (TuplePattern [BindingPattern (NamedPattern "a")]) 
case .y(true): ()
  // y(true) → EnumElementPattern "x" (TuplePattern [BoolPattern true])
case let y(b): ()
  // let y(b) → BindingPattern (EnumElementPattern "y" (TuplePattern [NamedPattern "b"]))
}
\endcode

Notice that \c TuplePattern does double duty; it plays the role of both
associated values for enum cases as well as for matching on actual tuples.
```

Reply:
Actually, I think these might be slightly wrong, due to the presence of `ParenPattern`, let me check...